stan/math/rev/core/autodiffstackstorage.hpp

```diff
@@ -9,40 +9,88 @@ namespace math {
 
 /**
  * Provides a thread_local singleton if needed. Read warnings below!
- * For performance reasons the singleton is a global static for the
- * case of no threading which is returned by a function. This design
- * should allow the compiler to apply necessary inlining to get
- * maximal performance. However, this design suffers from "the static
- * init order fiasco"[0].  Anywhere this is used, we must be
- * absolutely positive that it doesn't matter when the singleton will
- * get initialized relative to other static variables.  In exchange,
- * we get a more performant singleton pattern for the non-threading
- * case. In the threading case we use the defacto standard C++11
- * singleton pattern relying on a function wrapping a static local
- * variable. This standard pattern is expected to be well supported
- * by the major compilers (as its standard), but it does incur some
- * performance penalty.  There has been some discussion on this; see
- * [1] and [2] and the discussions those PRs link to as well.
+ * For performance reasons the singleton is a global static pointer
+ * which is stored as thread local (TLS) if STAN_THREADS is
+ * defined. The use of a pointer is motivated by performance reasons
+ * for the threading case. When a TLS is used then initialization with
+ * a constant expression is required for fast access to the TLS. As
+ * the AD storage struct is non-POD it must be initialized as a
+ * dynamic expression such that compilers will wrap any access to the
+ * TLS by a TLS wrapper function which causes a significant
+ * slow-down. A pointer to the AD storage instance can be initialized
+ * to a compile-time constant expression of nullptr. In this case the
+ * compiler avoids the use of a TLS wrapper function. Furthermore we
+ * use the __thread keyword on compilers which support it. The
+ * __thread keyword is a compiler-specific (gcc, clang, Intel)
+ * extension which requires initialization with a compile time
+ * constant expression. The C++11 keyword thread_local does allow for
+ * constant and dynamic initialization of the TLS. Thus, only the
+ * __thread keyword gurantees that constant initialization and it's
+ * implied speedup, is used.
  *
- * These are thread_local only if the user asks for it with
- * -DSTAN_THREADS. This is primarily because Apple clang compilers
- * before 2016 don't support thread_local and the additional
- * performance cost. We have proposed removing support for those[3],
- * and at that time we should evaluate the performance of a switch to
- * thread_local.  If there is no loss in performance, we can remove
- * this ifdef.
+ * The initialzation of the AD instance is handled by the lifetime of
+ * a AutodiffStackSingleton object. More specifically, the first
```

Comment:
This ownership point is important, but it's not as clearly linked to the code (no mention of `own_instance_` here) as I think it should be.

If I understand this correctly, the "first instance" if there are no threads is the global one defined in `init_chainablestack.hpp`? 

And if there are threads, this doc doesn't seem right... we don't need to call `init()` anymore?

Reply:
Do we really need to link to private members of the class in the doc if the behavior is described? Well, I added some refs now.

Your understanding is correct... but the one in `init_chainablestack.hpp` does not have to be the first one if the C++ user wants more control over things and links things together so that what is execute first changes, then the first instance will be another one.

I corrected the ill references to init.